### DISCUSSION TITLE
Redaction annotations

### DISCUSSION THREAD

**Discussion by Ougadougou75** [stephanrauh/ngx-extended-pdf-viewer#2015] [Q&A]
Hi!

Is it possible to use ngx-extended-viewer to create and display redaction annotations? Both programatically (given page numbers and coordinates of bounding boxes on the page) and by selecting text in the viewer.  

If anyone has any previous experience in this field and would like to share some insights and advice it would be greatly appreciated!

**Comment by stephanrauh** [comment]
You can use the text editor. It also has a programmatic API (see https://pdfviewer.net/extended-pdf-viewer/export-annotations). Does this suit your use-case?

The only problem with the programmatic API is that it's using an internal data structure of the base library which is subject to change without notice. Plus, the data structure is not documented.

**Reply by Ougadougou75** [reply]
Thank you for your quick response! I will give it a try! 

The internal data structure you are referring to, is that the "FreeTextEditor", that is used in the example your link pointed to? How did you find out that it existed and what it looked like?

Also, from your documentation I was not able to figure out how to reach the underlying pdf.js in order to call its API directly. Can it be done and if so - can you give me a hint?

Looking at the code of ngx-extended-pdf-viewer.service.ts I found this method, referenced by the example you pointed me to:

public addEditorAnnotation(serializedAnnotation: string | EditorAnnotation): void {
    const PDFViewerApplication: IPDFViewerApplication = (window as any).PDFViewerApplication;
    PDFViewerApplication.pdfViewer.addEditorAnnotation(serializedAnnotation);
  }

What is the PDFViewerApplication? Is that code you wrote or is it a part of the underlying pdf.js? Is it described somewhere so that I can better understand what the addEditorAnnotation() method does (and maybe also find other interesting methods)?
